### PR TITLE
APEXMALHAR-2042 #resolve #comment fixed race condition where state co…

### PR DESCRIPTION
**Pull request by ilooner**
…uld be cleared before data is transfered to a file.

### CODE REVIEW

**Comment by ilooner** [comment]
@chandnisingh please review.

File: library/src/main/java/org/apache/apex/malhar/lib/state/managed/IncrementalCheckpointManager.java [hunk 1]
```diff
@@ -208,6 +209,11 @@ public void setLatestExpiredTimeBucket(long timeBucket)
     latestExpiredTimeBucket.set(timeBucket);
   }
 
+  public long getLastTransferedWindow()
```
Comment (otterc) [inline library/src/main/java/org/apache/apex/malhar/lib/state/managed/IncrementalCheckpointManager.java]: please document this public method

File: library/src/test/java/org/apache/apex/malhar/lib/state/managed/ManagedStateImplTest.java [hunk 2]
```diff
@@ -147,9 +150,11 @@ public void testCommitted()
     Slice two = ManagedStateTestUtils.getSliceFor("2");
     commitHelper(one, two);
     Bucket.DefaultBucket defaultBucket = (Bucket.DefaultBucket)testMeta.managedState.getBucket(0);
-    Assert.assertEquals("value of one", one, defaultBucket.getCommittedData().get(one).getValue());
+    Assert.assertEquals("value of one",
+        one, defaultBucket.getCommittedData().firstEntry().getValue().get(one).getValue());
```
Comment (otterc) [inline library/src/test/java/org/apache/apex/malhar/lib/state/managed/ManagedStateImplTest.java]: this is less 120 lines. please don't break it into 2 lines

File: library/src/test/java/org/apache/apex/malhar/lib/state/managed/ManagedStateImplTest.java [hunk 3]
```diff
@@ -147,9 +150,11 @@ public void testCommitted()
...
 
-    Assert.assertNull("value of two", defaultBucket.getCommittedData().get(two));
+    Assert.assertNull("value of two",
+        defaultBucket.getCommittedData().firstEntry().getValue().get(two));
```
Comment (otterc) [inline library/src/test/java/org/apache/apex/malhar/lib/state/managed/ManagedStateImplTest.java]: Same as above

File: library/src/test/java/org/apache/apex/malhar/lib/state/managed/ManagedStateImplTest.java [hunk 4]
```diff
@@ -163,6 +168,45 @@ public void testAsyncGetFromCommitted() throws ExecutionException, InterruptedEx
     Assert.assertEquals("value of one", one, valFuture.get());
   }
 
+  /**
+   * Reproducing APEXMALHAR-2042
```
Comment (otterc) [inline library/src/test/java/org/apache/apex/malhar/lib/state/managed/ManagedStateImplTest.java]: Fix for APEXMALHAR-2042?

File: library/src/test/java/org/apache/apex/malhar/lib/state/managed/ManagedStateImplTest.java [hunk 5]
```diff
@@ -147,9 +150,9 @@ public void testCommitted()
     Slice two = ManagedStateTestUtils.getSliceFor("2");
     commitHelper(one, two);
     Bucket.DefaultBucket defaultBucket = (Bucket.DefaultBucket)testMeta.managedState.getBucket(0);
-    Assert.assertEquals("value of one", one, defaultBucket.getCommittedData().get(one).getValue());
-
-    Assert.assertNull("value of two", defaultBucket.getCommittedData().get(two));
+    Assert.assertEquals("value of one", one, defaultBucket.getCommittedData().firstEntry().getValue().get(one)
+        .getValue());
```
Comment (ilooner) [inline library/src/test/java/org/apache/apex/malhar/lib/state/managed/ManagedStateImplTest.java:154]: This one is longer than 120.

**Comment by ilooner** [comment]
@chandnisingh implemented comments

**Comment by ilooner** [comment]
@chandnisingh fixed typo

**Pull request merged**